### PR TITLE
SONARJAVA-1538 - CFG SE : nested statements in try catch end up with …

### PR DESCRIPTION
**Pull request by DidierBesset**
…wrong CFG

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/cfg/CFG.java [hunk 1]
```diff
@@ -765,21 +770,16 @@ private void buildTryStatement(TryStatementTree tryStatementTree) {
...
     build(tryStatementTree.block());
-    if(currentBlock.exitBlock != null && currentBlock.exitBlock.isFinallyBlock) {
-      for (Block catchBlock : catches) {
-        currentBlock.exitBlock.addSuccessor(catchBlock);
-      }
-    } else {
-      for (Block catchBlock : catches) {
-        currentBlock.addSuccessor(catchBlock);
-      }
+    if (currentBlock != beforeFinally) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/cfg/CFG.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: java-frontend/src/main/java/org/sonar/java/cfg/CFG.java [hunk 2]
```diff
@@ -788,6 +788,18 @@ private void buildTryStatement(TryStatementTree tryStatementTree) {
     }
   }
 
+  protected void linkToCatchBlocks(Block block, List<Block> catches) {
+    if (block.exitBlock != null && block.exitBlock.isFinallyBlock) {
+      for (Block catchBlock : catches) {
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/cfg/CFG.java]: Matter of style, but I would choose the block with the if and have only one loop for to add successors.

File: java-frontend/src/test/java/org/sonar/java/cfg/CFGTest.java [hunk 3]
```diff
@@ -1302,4 +1302,87 @@ public void method_reference() throws Exception {
...
+        "       performAction();" +
+        "    }" +
+        "} catch(Exception e) { foo();} bar(); }");
+    cfgChecker = checker(
+      block(
+        element(Tree.Kind.TRY_STATEMENT)).successors(3, 5),
+      block(
+        element(Tree.Kind.IDENTIFIER, "doSomething"),
+        element(Kind.METHOD_INVOCATION),
+        element(Tree.Kind.IDENTIFIER, "action")).terminator(Kind.IF_STATEMENT).successors(2, 3, 4),
```
Comment (benzonico) [inline java-frontend/src/test/java/org/sonar/java/cfg/CFGTest.java]: I am unsure to understand why the condition here can have successors n°3 (the catch block) seems that this is not the behaviour we expect ?

File: java-frontend/src/main/java/org/sonar/java/cfg/CFG.java [hunk 4]
```diff
@@ -254,6 +255,10 @@ private void prune() {
         block.id = id;
         id += 1;
       }
+      inactiveBlocks.removeAll(blocks);
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/cfg/CFG.java]: This part has nothing to do with the current ticket. Therefore it should not be included in this commit. A dedicated test case is also required.
